### ISSUE TITLE
Kokkos::atomic_add gives incorrect results for complex on HIP

### ISSUE THREAD

**Issue by brian-kelley** [kokkos/kokkos#3974]
``Kokkos::atomic_add`` with ``Kokkos::complex<double>`` produces incorrect results on HIP (tested on ROCm 4.0.0, MI25). The build was RelWithDebInfo. I can try other hardware or rocm versions and report back if that's helpful.

Changing the ExecSpace to Serial makes it print the expected value.

Full reproducer:
```
#include "Kokkos_Core.hpp"

using std::cout;

using ExecSpace = Kokkos::Experimental::HIP;
using MemSpace = Kokkos::Experimental::HIPSpace;

using Scalar = Kokkos::complex<double>;
using SingleView = Kokkos::View<Scalar, MemSpace>;
using HostSingleView = typename SingleView::HostMirror;

struct Functor
{
        Functor(const SingleView& v_) : v(v_) {}
        KOKKOS_INLINE_FUNCTION void operator()(int i) const
        {
                Scalar s(-i, i * 2);
                Kokkos::atomic_add(v.data(), s);
        }
        SingleView v;
};

int main()
{
        Kokkos::initialize();
        {
                SingleView v("v");
                Kokkos::parallel_for(
                                Kokkos::RangePolicy<ExecSpace>(0, 1000),
                                Functor(v));
                auto vhost = Kokkos::create_mirror_view(v);
                Kokkos::deep_copy(vhost, v);
                cout << "Actual result: " << vhost() << '\n';
                Scalar gold(0, 0);
                for(int i = 0; i < 1000; i++)
                {
                        Scalar s(-i, i * 2);
                        gold += s;
                }
                cout << "Correct result: " << gold << '\n';
        }
        Kokkos::finalize();
        return 0;
}
```
Example output (I ran it a bunch of times and the result was the same every time).
```
Actual result: (-439320,878640)
Correct result: (-499500,999000)
```

**Comment by Rombur** [comment]
Yes, this a known problem (see [here](https://github.com/kokkos/kokkos/blob/4b97a22ff7be7635116930bb97173058d6079202/core/unit_test/TestAtomics.hpp#L544-L557)). The code for the `atomic` should be replaced soon (see https://github.com/kokkos/kokkos/pull/3247). Since the code will be replaced and nobody has asked for this capability, I've not look into it too much. Is is something that's blocking for you?

**Comment by brian-kelley** [comment]
@Rombur I see, thanks. It's somewhat blocking KokkosKernels - in my latest build, this is the last thing needed to get all tests passing on HIP in our typical configuration. But there's nothing specific I'm working on that needs it right away.

**Comment by brian-kelley** [comment]
update: This reproducer still shows actual/correct results differing, with version 7b5ffcc322 (so with DESUL atomics).

**Comment by Rombur** [comment]
This should be fixed by #4159

**Comment by brian-kelley** [comment]
@Rombur Yep, I can confirm that this reproducer and all the KokkosKernels stuff using atomics are now working. Feel free to close whenever. Thanks!
